lib/_stream_writable.js

```diff
@@ -275,35 +275,42 @@ function validChunk(stream, state, chunk, cb) {
 
 Writable.prototype.write = function(chunk, encoding, cb) {
   const state = this._writableState;
-  var ret = false;
+
+  if (typeof encoding === 'function') {
+    cb = encoding;
+    encoding = null;
+  }
+
+  if (typeof cb !== 'function')
+    cb = nop;
+
+  if (state.ending || state.destroyed) {
+    const err = state.ending ? new ERR_STREAM_WRITE_AFTER_END() :
+      new ERR_STREAM_DESTROYED('write');
+    process.nextTick(cb, err);
+    errorOrDestroy(this, err);
+    return false;
+  }
```

Comment:
I think this should call `writeAfterEnd(this, cb)` in case of `state.ending`, and do something else with `state.destroyed`. I'm also not sure why you moved this check at the top of the file, before the `isBuf` assignment.
Note that the `if (state.ending)` clause at line 307 is still there, and that branch would never be touched.

Reply:
You are totally right. This also has some unnecessary changes. I'll fix.